### ISSUE TITLE
Suggestion: Comparison of weighted roll in README

### ISSUE THREAD

**Issue by 0racle** [titsuki/raku-Random-Choice#1]
Calling `.roll` on a Bag or Mix will do a weighted random selection. See [Baggy.roll](https://docs.perl6.org/type/Baggy#method_roll).

A simplified `choice` function might look something like this

```perl-6
sub choice(:$size where * ≥ 1 = 1, :@p where abs(1 - .sum) < 1e-3) {
    @p.pairs.Mix.roll($size)
}

say choice(:size(5), :p([0.1, 0.1, 0.1, 0.7]));
```

I'm not a stats guy, so I don't know if the distributions would be comparable to your module.

Perhaps your README could be updated to state any advantages or disadvantages over the above approach, particularly given that the above approve is faster than your module.

**Comment by titsuki** [comment]
@0racle Thanks for your advice!

> I don't know if the distributions would be comparable to your module.

It seems comparable to me since Mix.roll does the same task.
However `Mix.roll` (Mix.roll calls `QuantHash.MIX-ROLL` and `QuantHash.MIX-ROLL` is a main function) conducts this task by the inefficient `O(N)` algorithm whereas `Random::Choice` does it by the O(1) algorithm (I disregard the initialization cost since both of them costs O(N) time and it doesn't make performance difference theoretically):
https://github.com/rakudo/rakudo/blob/2018.12/src/core/Rakudo/QuantHash.pm6#L1053-L1071

So, Mix.roll becomes very slowly when it tries to roll a huge dice (i.e., large size `@p`).

Here is the benchmark result:

myrandom.p6:
```
use Bench;


my $bench = Bench.new;

my $repeat-times = 100;
my @p;
for ^1000 {
    @p.push: rand;
}
my $sum = @p.sum;
@p = @p.map(* / $sum);
my $size = 1000;

my $rc-code = sub {
    use Random::Choice;
    choice(:$size, :@p);
};

sub choice(:$size where * ≥ 1 = 1, :@p where abs(1 - .sum) < 1e-3) {
    @p.pairs.Mix.roll($size)
}

my $mix-code = sub {
    choice(:$size, :@p);
}

$bench.timethis($repeat-times, $rc-code, title => 'Random::Choice');
$bench.timethis($repeat-times, $mix-code, title => 'Mix');
```

```
$ perl6 myrandom.p6 
      time: 1.595 wallclock secs (1.636 usr 0.008 sys 1.644 cpu) @ 62.713/s (n=100)
               (warning: too few iterations for a reliable count)
      time: 70.175 wallclock secs (70.165 usr 0.012 sys 70.177 cpu) @ 1.425/s (n=100)
               (warning: too few iterations for a reliable count)

```

>Perhaps your README could be updated to state any advantages or disadvantages over the above approach, particularly given that the above approve is faster than your module.

Good idea! I'll write it sometime soon. (But maybe rakudo itself should employ a faster algorithm...)

**Comment by titsuki** [comment]
Fixed in version 0.0.4
